### DISCUSSION TITLE
Proposed/Implemented New Packages for `core`

### DISCUSSION THREAD

**Discussion by gingerBill** [odin-lang/Odin#978] [Proposals]
Proposed New Packages for `core`

* [ ] `archive/tar`
* [ ] `archive/zip`
* [x] `bufio` buffered I/O streams
* [x] `bytes` []byte version of `strings`
* [x] `container` - more general container types
* [x] `c/token` - tokenizer for C with an inline preprocessor
* [ ] `c/parser` - parser for C
* [x] `compress/gzip`
* [ ] `compress/lzw`
* [ ] `compress/lz4`
* [x] `compress/zlib`
* [x] `encoding/csv`
* [x] `encoding/ini`
* [ ] `encoding/punycode` (maybe?)
* [ ] `encoding/toml` (maybe?)
* [x] `encoding/xml`
* [x] `io` generic data streams
* [x] `math/big` - arbitrary-precision arithmetic (i.e. big numbers) (integers, rationals, floating-points)
* [x] `net` - portable interface for network I/O (TCP/IP, UDP, Unix sockets, etc)
* [ ] `net/http` - HTTP client and server implementations
* [ ] `net/smtp` - implementation of Simple Mail Transfer Protocol
* [ ] `odin/printer` - implements printing of AST nodes, useful for automatic formatting
* [ ] `odin/checker` - odin semantic checker (compiler frontend)
* [ ] `regexp` - regular expression matching, probably based on https://swtch.com/~rsc/regexp/regexp1.html
* [ ] `terminal` - package to control the _terminal/console_, allowing for colouring and cursor moving.
* [x] `text/scanner` - general scanner for work on Odin/C like strings
* [ ] `text/template` - data-driven template generation for textual output
* [ ] `time/tzdata` - embedded copy of the timezone database

**Comment by thebirk** [comment]
I wrote a basic tar and zip reader a while back, might serve as a, somewhat, useful start for `archive/tar` and `archive/zip`. At the very least it has some structs 😄 

Tar reader: https://gist.github.com/thebirk/4d78d0884b963ab7175e708a6933b681
Zip reader: https://gist.github.com/thebirk/bd478a8d0912f42b3e58cdae178dddbe

**Comment by gingerBill** [comment]
`package container` probably needs to be completely redone and rethought out. It might make more sense to have a sub-package for each data structure rather than have them in one big package.

And my original idea to reimplementations of a map and dynamic array was not a good idea, especially since it was not idiomatic.

**Comment by Kelimion** [comment]
`compress/gzip` and `compress/zlib` have their decompression routines done. Compression is underway.

**Comment by kleeon** [comment]
I wrote a package for reading and writing zip files loosely based on Go's `archive/zip`: https://github.com/kleeon/odin-zip

I'm not sure if it's good enough to be included into `core`, but it supports pretty much everything you would need and is pretty fast.

**Reply by Kelimion** [reply]
`core:thing` cannot rely on `vendor:thing`, sorry. I'm sure you've done a great job otherwise, but we'll need zlib compression in `core`.

**Reply by kleeon** [reply]
Fair enough. I wanted to use `compress/zlib`, but couldn't because of no DEFLATE support. Also I wanted to support streaming and `compress/zlib` does not have that either unfortunately.

**Reply by Kelimion** [reply]
Deflate and streaming are both planned for the relative near term, at which point we'll revisit `archive/zip`, `archive/tar`, etc.

**Reply by kleeon** [reply]
Cool, looking forward to that.

**Comment by laytan** [comment]
I wrote an INI parser a few months ago (https://github.com/laytan/odin-ini-parser) and just came across this issue, would this be a good fit for encoding/ini, if so, I can PR it.
It is only parsing at the moment though, I can add encoding pretty easily I think.

I am also working on a net/http like package (https://github.com/laytan/odin-http) which currently has a HTTP/1.1 server implemented.

**Comment by Beefster09** [comment]
I think I'm going to try taking a crack at a regexp package.

Anything I should keep in mind?

**Reply by Kelimion** [reply]
`@mjolnir` on Discord was working on that a couple months ago. Maybe coordinate?

**Comment by Beefster09** [comment]
Are there any style guidelines for writing packages for core?

I've noticed they typically take an allocator as an optional argument, and that typically defaults to `context.allocator` (not `temp_allocator`), but I'm not sure if that's a hard rule.

I will probably also take a crack at the `terminal` package since I have done a fair bit of work with ANSI escapes in the past and can at least implement colors if nothing else.

**Reply by OnlyXuul** [reply]
I just found this thread.
I created an ANSI printing library called "afmt" to be used from "shared" folder. It's posted in showcase section of main Odin website forums. If anyone is still working on adding ANSI features to Odin core, please feel free to either: take what you want from that library, or consider it temporary until those features are implemented. I do not wish to de-rail any current efforts.

As a side note, I’ve got several ANSI based procedures for controlling the cursor/terminal that I’ve considered adding to afmt. I’d be happy to share if anyone is wanting to see those for the above proposed features. Let me know.

So far I have:

    scroll_up
    scroll_down
    cursor_return
    cursor_hide
    cursor_show
    cursor_save
    cursor_restore
    cursor_place
    cursor_bottom
    cursor_position
    get_terminal_size

**Comment by TiltingCoder** [comment]
what about a package for window, audio and input similar to thread wich abstract all the thread os api.
